pkg/chunk/chunk_store.go

```diff
@@ -482,3 +487,126 @@ func (c *store) convertChunkIDsToChunks(ctx context.Context, userID string, chun
 
 	return chunkSet, nil
 }
+
+func (c *store) DeleteChunk(ctx context.Context, from, through model.Time, userID, chunkID string, metric labels.Labels, partiallyDeletedInterval *model.Interval) error {
+	metricName := metric.Get(model.MetricNameLabel)
+	if metricName == "" {
+		return ErrMetricNameLabelMissing
+	}
+
+	chunkWriteEntries, err := c.schema.GetWriteEntries(from, through, userID, string(metricName), metric, chunkID)
+	if err != nil {
+		return errors.Wrapf(err, "when getting index entries to delete for chunkID=%s", chunkID)
+	}
+
+	return c.deleteChunk(ctx, userID, chunkID, metric, chunkWriteEntries, partiallyDeletedInterval, func(chunk Chunk) error {
+		return c.PutOne(ctx, chunk.From, chunk.Through, chunk)
+	})
+}
+
+func (c *store) deleteChunk(ctx context.Context,
+	userID string,
+	chunkID string,
+	metric labels.Labels,
+	chunkWriteEntries []IndexEntry,
+	partiallyDeletedInterval *model.Interval,
+	putChunkFunc func(chunk Chunk) error) error {
+
+	metricName := metric.Get(model.MetricNameLabel)
+	if metricName == "" {
+		return ErrMetricNameLabelMissing
+	}
+
+	// if chunk is partially deleted, fetch it, slice non-deleted portion and put it to store before deleting original chunk
+	if partiallyDeletedInterval != nil {
+		err := c.reboundChunk(ctx, userID, chunkID, *partiallyDeletedInterval, putChunkFunc)
```

Comment:
In case the error `ErrParialDeleteChunkNoOverlap`, shouldn't we ignore it and `return nil`? It means there's nothing to do but it's also OK.

Reply:
I wanted to return an error to tell the caller that it is doing something wrong. 
A caller should never try to delete chunk partially with an interval which is out of its range.
What do you think?

Reply:
Good point. Sounds good as is then.